### PR TITLE
Remove unqualified fields from CFI.

### PR DESCRIPTION
**Pull request by luqmana**

### CODE REVIEW

**Comment by wmdietl** [comment]
The SPARTA test case is failing:
https://travis-ci.org/opprop/checker-framework-inference/jobs/164156164
Can you have a look why this is failing in this branch?

**Comment by topnessman** [comment]
The second commit seems like a hack for me.
There is a sentence in documentation of DefaultQualifierInHierarchy:

> Each type qualifier hierarchy may have at most one qualifier marked as
 * {@code DefaultQualifierInHierarchy}.

You annotated both @Sink and @Source. So it seems a bit problematic for me. What do you guys think?

**Comment by wmdietl** [comment]
Sink and Source are two different hierarchies, so that might be ok.
However, all of the defaulting is done imperatively in https://github.com/opprop/checker-framework-inference/blob/3dad429379abb45eff0327f15e9f53752f5ce1ec/src/sparta/checkers/SimpleFlowAnnotatedTypeFactory.java#L210
It would be better to remove the `@DefaultQualifierInHierarchy` again and make sure that all cases are covered in the factory.

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 1]
```diff
@@ -307,7 +273,7 @@ public AnnotationMirror getTopAnnotation(final AnnotationMirror am) {
             return varAnnot;
         } //else
 
-        return unqualified;
+        return inferenceMain.getRealTypeFactory().getQualifierHierarchy().getTopAnnotations().iterator().next();
```
Comment (wmdietl) [inline src/checkers/inference/InferenceQualifierHierarchy.java]: Is this branch needed? From other changes it looks like only VarAnnot subtypes are expected.

File: src/checkers/inference/VariableAnnotator.java [hunk 2]
```diff
@@ -573,18 +568,18 @@ private VariableSlot createEquivalentSlotConstraints(AnnotatedTypeMirror atm, Tr
...
         AnnotationMirror existinVar = atm.getAnnotationInHierarchy(varAnnot);
         if (!atm.getAnnotations().isEmpty()) {
-            realQualifier = atm.getAnnotationInHierarchy(unqualified);
-
-            if (!InferenceQualifierHierarchy.isUnqualified(realQualifier)) {
-                constantSlot = slotManager.getSlot(realQualifier);
+            realQualifier = atm.getAnnotationInHierarchy(realTop);
+            if (realQualifier == null) {
+                atm.addAnnotation(realTop);
+                realQualifier = atm.getAnnotationInHierarchy(realTop);
```
Comment (wmdietl) [inline src/checkers/inference/VariableAnnotator.java]: Won't this just return realTop?

File: src/checkers/inference/InferenceQualifierHierarchy.java (hunk 1 above)
Comment (wmdietl) [inline src/checkers/inference/InferenceQualifierHierarchy.java]: Yes, but on "new" line 58 above you have `"There should be only 1 top qualifier ( checkers.inference.qual.VarAnnot ).\n"`
This method determines the top qualifier, so shouldn't the answer also be `VarAnnot` if `am` is from the real hierarchy?

Comment (luqmana) [inline src/checkers/inference/InferenceQualifierHierarchy.java]: Ok, updated to return `this.getTopAnnotations()` which already filters out the non-VarAnnot ones in the `finish` method (line 73-91).

**Pull request merged**